### PR TITLE
Set GVK when getting the object from the cache.

### LINKED ISSUE #202: TypeMeta : Kind + ApiVersion are empty when the object is processed the first time
## Bug

Original ticket : https://github.com/operator-framework/operator-sdk/issues/727

When I create a `Component`'s custom resource, the component is well created and the MetaData well defined (Kind, ApiVersion)

```
oc get cp/fruit-client -o yaml                
apiVersion: component.k8s.io/v1alpha1
kind: Component
metadata:
  name: fruit-client
  namespace: my-spring-app
spec:
  deployment: innerloop
  exposeService: true
  runtime: springboot
  version: 1.5.16
```

but when the reconcile's function of the controller fetch the Component, then the Kind, ApiVersion fields are empty

What is the root cause of this issue ?

screenshot 2018-11-09 12 27 43

## Environment

```
okd : 3.11 (= kubernetes v1.11)
operator-sdk : v0.1.1+git
```

## Project

Code of the API : https://github.com/snowdrop/component-operator/tree/operator-0.1.1/pkg/apis
Code of the controller : https://github.com/snowdrop/component-operator/blob/operator-0.1.1/pkg/controller/component/handler.go#L94

## Hypothesis

As described [here](https://github.com/operator-framework/operator-sdk/issues/727#issuecomment-437343992) I suspect that we have a `deepCopy` issue the first time that we get the object from the cache

### PR DESCRIPTION
**Pull request by shawn-hurley**
* Fixing up a TODO around DeepCopyInto

fixes #202

### CODE REVIEW

**Comment by cmoulliard** [comment]
I have applied the patch to test it with my project and when an existing resource is updated now, I get this error during the step to create the resource

```
ERRO[0021] deploymentconfigs.apps.openshift.io "fruit-client" is invalid: metadata.resourceVersion: Invalid value: 0x0: must be specified for an update
```

**Comment by cmoulliard** [comment]
Forget the error reported in my comment as it is related to a bug into my own code. Sorry.

File: pkg/cache/internal/cache_reader.go [hunk 1]
```diff
@@ -73,14 +73,20 @@ func (c *CacheReader) Get(_ context.Context, key client.ObjectKey, out runtime.O
...
-	// Copy the value of the item in the cache to the returned value
-	// TODO(directxman12): this is a terrible hack, pls fix (we should have deepcopyinto)
-	outVal := reflect.ValueOf(out)
-	objVal := reflect.ValueOf(obj)
-	if !objVal.Type().AssignableTo(outVal.Type()) {
-		return fmt.Errorf("cache had type %s, but %s was asked for", objVal.Type(), outVal.Type())
+	oType := reflect.TypeOf(obj)
+	_, ok := oType.MethodByName("DeepCopyInto")
+	if ok {
+		reflect.ValueOf(obj).MethodByName("DeepCopyInto").Call([]reflect.Value{reflect.ValueOf(out)})
```
Comment (pwittrock) [inline pkg/cache/internal/cache_reader.go]: Do we need to check for an error here?

File: pkg/cache/internal/cache_reader.go [hunk 2]
```diff
@@ -73,14 +73,20 @@ func (c *CacheReader) Get(_ context.Context, key client.ObjectKey, out runtime.O
 	// TODO(directxman12): revisit the decision to always deepcopy
 	obj = obj.(runtime.Object).DeepCopyObject()
 
-	// Copy the value of the item in the cache to the returned value
-	// TODO(directxman12): this is a terrible hack, pls fix (we should have deepcopyinto)
-	outVal := reflect.ValueOf(out)
-	objVal := reflect.ValueOf(obj)
-	if !objVal.Type().AssignableTo(outVal.Type()) {
-		return fmt.Errorf("cache had type %s, but %s was asked for", objVal.Type(), outVal.Type())
+	oType := reflect.TypeOf(obj)
```
Comment (pwittrock) [inline pkg/cache/internal/cache_reader.go]: We could probably do this without reflection if the deepcopy code generated this function.  That should be relatively straightforward to add.  Would you add a TODO for that?

File: pkg/cache/internal/cache_reader.go [hunk 3]
```diff
@@ -73,14 +73,20 @@ func (c *CacheReader) Get(_ context.Context, key client.ObjectKey, out runtime.O
...
+	_, ok := oType.MethodByName("DeepCopyInto")
+	if ok {
```
Comment (pwittrock) [inline pkg/cache/internal/cache_reader.go]: Would you pull this logic into its own function and unit test it?  Verify both code paths work?

File: pkg/cache/internal/cache_reader.go [hunk 4]
```diff
@@ -73,14 +73,20 @@ func (c *CacheReader) Get(_ context.Context, key client.ObjectKey, out runtime.O
...
+		// Copy the value of the item in the cache to the returned value
+		outVal := reflect.ValueOf(out)
+		objVal := reflect.ValueOf(obj)
+		if !objVal.Type().AssignableTo(outVal.Type()) {
+			return fmt.Errorf("cache had type %s, but %s was asked for", objVal.Type(), outVal.Type())
+		}
+		reflect.Indirect(outVal).Set(reflect.Indirect(objVal))
 	}
-	reflect.Indirect(outVal).Set(reflect.Indirect(objVal))
+	out.GetObjectKind().SetGroupVersionKind(c.groupVersionKind)
```
Comment (pwittrock) [inline pkg/cache/internal/cache_reader.go]: Would update any tests to verify this is being populated now.

File: pkg/cache/internal/cache_reader.go (hunk 1 above)
Comment (shawn-hurley) [reply pkg/cache/internal/cache_reader.go]: Actually, I don't think we do, 

https://github.com/kubernetes/api/blob/master/admission/v1beta1/zz_generated.deepcopy.go#L29

Does that sound correct?

File: pkg/cache/internal/cache_reader.go (hunk 2 above)
Comment (DirectXMan12) [reply pkg/cache/internal/cache_reader.go]: Is there not an interface for `DeepCopyInto`?

Comment (shawn-hurley) [reply pkg/cache/internal/cache_reader.go]: Hey, guys so I don't think that we can use an interface based on how generation works today. 
https://github.com/kubernetes/api/blob/master/core/v1/zz_generated.deepcopy.go#L30

If you notice that the `DeepCopyInto` does not use a `runtime.Object` as its input, it instead uses the type. I believe this means that we can not make a generic interface for `DeepCopyInto`. 

Am I missing something?

**Comment by shawn-hurley** [comment]
Closing in favor of two PRs

File: pkg/cache/internal/cache_reader.go (hunk 2 above)
Comment (pwittrock) [reply pkg/cache/internal/cache_reader.go]: Sorry, I was suggesting as a follow up we make a function that takes a runtime.Object and checks its type.  It should be trivial to wrap the DeepCopyInto function in another function that takes runtime.Object and does a type check.

Again, that would be a follow up and something we could put in the 1.14 version

File: pkg/cache/internal/cache_reader.go (hunk 3 above)
Comment (pwittrock) [reply pkg/cache/internal/cache_reader.go]: The first one.

**Comment by cmoulliard** [comment]
> Closing in favor of two PRs

What are the 2 PRs you mentioned ? Have they been merged and released ?

**Pull request closed without merging**